libraries/AP_ExternalAHRS/AP_ExternalAHRS_AnelloX3.cpp

```diff
@@ -0,0 +1,441 @@
+        };
+        AP::ins().handle_external(ins);
+    }
+
+#if AP_COMPASS_EXTERNALAHRS_ENABLED
+    {
+        AP_ExternalAHRS::mag_data_message_t mag {
+            field: imu_data.mag
+        };
+        AP::compass().handle_external(mag);
+    }
+#endif
+
+}
+
+int8_t AP_ExternalAHRS_AnelloX3::get_port(void) const
+{
+    if (!uart) {
+        return -1;
+    }
+    return port_num;
+};
+
+const char* AP_ExternalAHRS_AnelloX3::get_name() const
+{
+    return "Anello X3";
+}
+
+bool AP_ExternalAHRS_AnelloX3::healthy(void) const
+{
+    uint32_t now = AP_HAL::millis();
+
+    // unhealthy if packet delayed or fusion status flags are raised
+    return (now - last_imu_pkt < 40 && imu_data.fusion_status_x == 0 &&
+            imu_data.fusion_status_y == 0 && imu_data.fusion_status_z == 0);
+}
+
+bool AP_ExternalAHRS_AnelloX3::initialised(void) const
+{
+    return last_imu_pkt != 0;
+}
+
+bool AP_ExternalAHRS_AnelloX3::pre_arm_check(char *failure_msg, uint8_t failure_msg_len) const
+{
+    if (!healthy()) {
+        hal.util->snprintf(failure_msg, failure_msg_len, "Anello X3 unhealthy");
+        return false;
+    }
+    return true;
+}
+
+void AP_ExternalAHRS_AnelloX3::get_filter_status(nav_filter_status &status) const
+{
+    // no filter status info from sensor 
+}
+
+bool AP_ExternalAHRS_AnelloX3::get_variances(float &velVar, float &posVar, float &hgtVar, Vector3f &magVar, float &tasVar) const
+{
+    // no variance data from sensor
+    return false;
+}
+
+
+uint8_t AP_ExternalAHRS_AnelloX3::num_gps_sensors(void) const
+{
+    // system does not host GPS sensors
+    return 0;
+}
+
+// checksum verification
+bool AP_ExternalAHRS_AnelloX3::valid_packet(const AnelloX3_Packet & packet)
+{
+    uint8_t checksum_one = 0;
+    uint8_t checksum_two = 0;
+
+    for (int i = 2; i < 4; i++) {
+        checksum_one += packet.header[i];
+        checksum_two += checksum_one;
+    }
+
+    for (int i = 0; i < packet.payload_length(); i++) {
+        checksum_one += packet.payload[i];
+        checksum_two += checksum_one;
+    }
+
+    return packet.checksum[0] == checksum_one && packet.checksum[1] == checksum_two;
+}
+
+// directs differnt message types to resp parsers (currently, only IMU)
+AP_ExternalAHRS_AnelloX3::DescriptorSet AP_ExternalAHRS_AnelloX3::handle_packet(const AnelloX3_Packet& packet)
+{
+    const DescriptorSet descriptor = packet.descriptor_set();
+    switch (descriptor) {
+    case DescriptorSet::IMUData:
+        handle_imu(packet);
+        break;
+    }
+    return descriptor;
+}
+
+// unpacking function for the X3 IMU messages
+void AP_ExternalAHRS_AnelloX3::handle_imu(const AnelloX3_Packet& packet)
+{
+    // keep track of last recv packet
+    last_imu_pkt = AP_HAL::millis(); 
+
+    // struct for holding raw binary data with endian conversion
+    struct AnelloX3_BinaryPayload bin_payload;
+
+    // state variables for unpacking
+    int i = 0;
+
+    // go through payload chunk-by-chunk
```

Comment:
This would be better done by casting packet to a structure and extracting the values from said structure.

The extra intermediate step of packet -> bin_payload -> imu_data doesn't seem warranted.